cmd/clusterctl/client/cluster/topology_test.go

```diff
@@ -0,0 +1,693 @@
+	type args struct {
+		in *DryRunInput
+	}
+	type item struct {
+		kind       string
+		namespace  string
+		namePrefix string
+	}
+	type out struct {
+		affectedClusters       []client.ObjectKey
+		affectedClusterClasses []client.ObjectKey
+		created                []item
+		modified               []item
+		deleted                []item
+	}
+	tests := []struct {
+		name            string
+		existingObjects []*unstructured.Unstructured
+		args            args
+		want            out
+		wantErr         bool
+	}{
+		{
+			name: "Input with new ClusterClass and new Cluster",
+			args: args{
+				in: func() *DryRunInput {
+					objs, err := utilyaml.ToUnstructured(newClusterClassAndClusterYAML)
+					if err != nil {
+						panic(err)
+					}
+					return &DryRunInput{
+						Objs: convertToPtrSlice(objs),
+					}
+				}(),
+			},
+			want: out{
+				created: []item{
+					{kind: "DockerCluster", namespace: "default", namePrefix: "my-cluster-"},
+					{kind: "DockerMachineTemplate", namespace: "default", namePrefix: "my-cluster-md-0-"},
+					{kind: "DockerMachineTemplate", namespace: "default", namePrefix: "my-cluster-md-1-"},
+					{kind: "DockerMachineTemplate", namespace: "default", namePrefix: "my-cluster-control-plane-"},
+					{kind: "KubeadmConfigTemplate", namespace: "default", namePrefix: "my-cluster-md-0-bootstrap-"},
+					{kind: "KubeadmConfigTemplate", namespace: "default", namePrefix: "my-cluster-md-1-bootstrap-"},
+					{kind: "KubeadmControlPlane", namespace: "default", namePrefix: "my-cluster-"},
+					{kind: "MachineDeployment", namespace: "default", namePrefix: "my-cluster-md-0-"},
+					{kind: "MachineDeployment", namespace: "default", namePrefix: "my-cluster-md-1-"},
+				},
+				modified: []item{
+					{kind: "Cluster", namespace: "default", namePrefix: "my-cluster"},
+				},
+				affectedClusters: func() []client.ObjectKey {
+					cluster := client.ObjectKey{}
+					cluster.Namespace = "default"
+					cluster.Name = "my-cluster"
+					return []client.ObjectKey{cluster}
+				}(),
+				affectedClusterClasses: func() []client.ObjectKey {
+					cc := client.ObjectKey{}
+					cc.Namespace = "default"
+					cc.Name = "my-cluster-class"
+					return []client.ObjectKey{cc}
+				}(),
+			},
+			wantErr: false,
+		},
+		{
+			name: "Modifying an existing Cluster",
+			existingObjects: func() []*unstructured.Unstructured {
+				objs, err := utilyaml.ToUnstructured(existingClusterClassAndClusterYAML)
+				if err != nil {
+					panic(err)
+				}
+				return convertToPtrSlice(objs)
+			}(),
+			args: args{
+				in: func() *DryRunInput {
+					objs, err := utilyaml.ToUnstructured(modifiedClusterYAML)
+					if err != nil {
+						panic(err)
+					}
+					return &DryRunInput{
+						Objs: convertToPtrSlice(objs),
+					}
+				}(),
+			},
+			want: out{
+				modified: []item{
+					{kind: "KubeadmControlPlane", namespace: "default", namePrefix: "my-cluster-"},
+				},
+			},
+			wantErr: false,
+		},
+	}
```

Comment:
I think we need a couple of other tests verifying ataffectedClusterClasses variants and affected cluster variants